### ISSUE TITLE
panic while updating an azurerm_linux_virtual_machine_scale_set

### ISSUE THREAD

**Issue by beparmentier** [terraform-providers/terraform-provider-azurerm#12797]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

**terraform 1.0.3**

```hcl
Initializing provider plugins...
- Installing hashicorp/random v3.1.0...
- Installed hashicorp/random v3.1.0 (signed by HashiCorp)
- Installing hashicorp/azurerm v2.70.0...
- Installed hashicorp/azurerm v2.70.0 (signed by HashiCorp)
```

### Affected Resource(s)

* `azurerm_2.70.0`

### Terraform Configuration Files

```hcl
resource "azurerm_linux_virtual_machine_scale_set" "scalesetsemarchy" {
  for_each                        = local.semarchy_instance
  name                            = each.value.scaleset_name
  resource_group_name             = azurerm_resource_group.rgenv[each.key].name
  location                        = azurerm_resource_group.rgenv[each.key].location
  instances                       = 1
  sku                             = each.value.sku
  admin_username                  = each.value.adminusername
  admin_password                  = random_password.semarchypasswd[each.key].result
  disable_password_authentication = each.value.disable_password_authentication
  zone_balance                    = each.value.zone_balance
  upgrade_mode                    = each.value.upgrade_mode
  custom_data                     = base64encode(local.templatecloudinitconfigsemarchy)
  zones                           = each.value.zones
  source_image_id                 = var.semarchy_image

  os_disk {
    storage_account_type = each.value.storage_account_type
    disk_size_gb         = each.value.disk_size_gb
    caching              = each.value.caching
  }

  boot_diagnostics {
    storage_account_uri = azurerm_storage_account.storsemarchy[each.key].primary_blob_endpoint
  }

  automatic_instance_repair {
    enabled = true
  }

  automatic_os_upgrade_policy {
    disable_automatic_rollback  = false
    enable_automatic_os_upgrade = true
  }

  network_interface {
    name                          = "nic"
    primary                       = true
    enable_accelerated_networking = true
    ip_configuration {
      name                                   = "internal"
      primary                                = true
      subnet_id                              = azurerm_subnet.this["${each.key}-${local.env}-${local.semarchy_layer}"].id
      load_balancer_backend_address_pool_ids = [azurerm_lb_backend_address_pool.backendpool[each.key].id]
    }
  }

  tags = local.global_tags
}
```

### Panic Output

```hcl
│ Error: Plugin did not respond
│ 
│   with azurerm_linux_virtual_machine_scale_set.scaleset[MASKED]["westeurope"],
│   on 05_[MASKED].tf line 29, in resource "azurerm_linux_virtual_machine_scale_set" "scaleset[MASKED]":
│   29: resource "azurerm_linux_virtual_machine_scale_set" "scaleset[MASKED]" {
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵
Stack trace from the terraform-provider-azurerm_v2.70.0_x5 plugin:
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x4910b4d]
...
```

### Expected Behaviour

Just applying the expected plan [terraform plan](https://gist.github.com/beparmentier/287e828509ffd0b551f79a6958cd1c66)

### Actual Behaviour

[terraform panic](https://gist.github.com/beparmentier/59be18203e4d337d88f33e3d19092f63)

terraform crash with a panic

### Steps to Reproduce

I don't really know...

I don't get any issue while creating the scaleset.

I just update the `azurerm_linux_virtual_machine_scale_set` to add

```hcl
  automatic_instance_repair {
    enabled = true
  }

  automatic_os_upgrade_policy {
    disable_automatic_rollback  = false
    enable_automatic_os_upgrade = true
  }
```

and I get this error

I also got this issue while updating the `source_image_id`.

### Important Factoids

Running terraform in Azure West Europe

### References

* #0000

**Comment by tombuildsstuff** [comment]
Stack trace:

```

github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/compute.virtualMachineScaleSetUpdateMetaData.performUpdate(0x10101, 0xc000170780, 0xc001e25e60, 0xc00134ce70, 0x0, 0xc001e26f00, 0x0, 0xc00134cd38, 0x0, 0xc0015dfe80, ...)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/compute/virtual_machine_scale_set_update.go:52 +0x14d
github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/compute.resourceLinuxVirtualMachineScaleSetUpdate(0xc001c05980, 0x57787e0, 0xc000849500, 0x0, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/compute/linux_virtual_machine_scale_set_resource.go:842 +0xc30
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).update(0xc000691880, 0x670a628, 0xc001be5bc0, 0xc001c05980, 0x57787e0, 0xc000849500, 0x0, 0x0, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/resource.go:352 +0x1ee
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc000691880, 0x670a628, 0xc001be5bc0, 0xc001194460, 0xc001028360, 0x57787e0, 0xc000849500, 0x0, 0x0, 0x0, ...)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/resource.go:464 +0x390
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc001160330, 0x670a628, 0xc001be5bc0, 0xc001bc9040, 0xc001be5bc0, 0x5c595a0, 0xc0024c4500)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/grpc_provider.go:955 +0x8ef
github.com/hashicorp/terraform-plugin-go/tfprotov5/server.(*server).ApplyResourceChange(0xc000e7e5c0, 0x670a6d0, 0xc001be5bc0, 0xc001194230, 0xc000e7e5c0, 0xc0024c45a0, 0xc000d53ba0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform-plugin-go/tfprotov5/server/server.go:332 +0xb5
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler(0x5c595a0, 0xc000e7e5c0, 0x670a6d0, 0xc0024c45a0, 0xc001b37740, 0x0, 0x670a6d0, 0xc0024c45a0, 0xc00174ea00, 0x23c5)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:380 +0x214
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000cbe380, 0x67565b8, 0xc000e8ef00, 0xc0024bf320, 0xc00122a4e0, 0x9cccf80, 0x0, 0x0, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc/server.go:1292 +0x52b
google.golang.org/grpc.(*Server).handleStream(0xc000cbe380, 0x67565b8, 0xc000e8ef00, 0xc0024bf320, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc/server.go:1617 +0xd0c
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc00082cd60, 0xc000cbe380, 0x67565b8, 0xc000e8ef00, 0xc0024bf320)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc/server.go:940 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc/server.go:938 +0x1fd
```

**Comment by beparmentier** [comment]
Hello,

I made few more tests and I tried to push my changes individually.

Seems there is a problem while using `source_image_id` and `automatic_os_upgrade_policy` when the source image doesn't provide from Azure Image Gallery.

```hcl
│ Error: Error creating Linux Virtual Machine Scale Set "vmss-npr-int-mdm-[MASKED]" (Resource Group "rg_npr_int_mdm_integ_we"): compute.VirtualMachineScaleSetsClient#CreateOrUpdate: Failure sending request: StatusCode=0 -- Original Error: Code="BadRequest" Message="virtualMachineScaleSets with Automatic OS Upgrade enabled must use Platform or Gallery Image with version set to 'latest'."
```

And according to the document `automatic_instance_repair`  needs a `health_probe_id` or an ApplicationHealth extension. But when the extension is described in a `azurerm_virtual_machine_scale_set_extension` it doesn't works.

```hcl
│ Error: Error creating Linux Virtual Machine Scale Set "vmss-npr-int-mdm-[MASKED]" (Resource Group "rg_npr_int_mdm_integ_we"): compute.VirtualMachineScaleSetsClient#CreateOrUpdate: Failure sending request: StatusCode=0 -- Original Error: Code="BadRequest" Message="Automatic repairs not supported for this Virtual Machine Scale Set because a health probe or health extension was not provided."
```

Adding an `extension` block in `azurerm_linux_virtual_machine_scale_set` solves the issue.

**Comment by jackofallops** [comment]
Hi @beparmentier - In this instance the panic was a nil dereference in the trying to check the value for the image in use in this line
```
virtual_machine_scale_set_update.go:52
```
The fix here guards against this access. Your use case should not have been able to crash the provider,  even if the configuration was incomplete as I think you are suggesting above.  Many thanks for your additional investigation and detail. This fix will go out in this week's release of 2.71.0
